RecoParticleFlow/PFClusterProducerAlpaka/plugins/alpaka/PFClusterProducerKernel.dev.cc

```diff
@@ -0,0 +1,2224 @@
+    Position4& seedPos = alpaka::declareSharedVar<Position4, __COUNTER__>(acc);
+    bool& notDone = alpaka::declareSharedVar<bool, __COUNTER__>(acc);
+    bool& debug = alpaka::declareSharedVar<bool, __COUNTER__>(acc);
+    if (tid == 0) {
+      i = clusteringVars[clusteringVars[topoId].topoSeedOffsets()].topoSeedList();  // i is the seed rechit index
+      nRHOther = nRHTopo - 1;                                                       // number of non-seed rechits
+      seedPos = Position4{pfRecHits[i].x(), pfRecHits[i].y(), pfRecHits[i].z(), 1.};
+      clusterPos = seedPos;  // Initial cluster position is just the seed
+      prevClusterPos = seedPos;
+      seedEnergy = pfRecHits[i].energy();
+      clusterEnergy = seedEnergy;
+      tol = pfClusParams.stoppingTolerance();  // stopping tolerance * tolerance scaling
+
+      if (pfRecHits[i].layer() == PFLayer::HCAL_BARREL1)
+        rhENormInv = pfClusParams.recHitEnergyNormInvEB_vec()[pfRecHits[i].depth() - 1];
+      else if (pfRecHits[i].layer() == PFLayer::HCAL_ENDCAP)
+        rhENormInv = pfClusParams.recHitEnergyNormInvEE_vec()[pfRecHits[i].depth() - 1];
+      else {
+        rhENormInv = 0.;
+        printf("Rechit %d has invalid layer %d!\n", i, pfRecHits[i].layer());
+      }
+
+      iter = 0;
+      notDone = true;
+      debug = false;
+    }
+    alpaka::syncBlockThreads(acc);
+
+    int j = -1;  // j is the rechit index for this thread
+    int rhFracOffset = -1;
+    Position4 rhPos;
+    float rhEnergy = -1., rhPosNorm = -1.;
+
+    if (tid < nRHOther) {
+      rhFracOffset =
+          clusteringVars[i].seedFracOffsets() + tid + 1;  // Offset for this rechit in pcrhfrac, pcrhfracidx arrays
+      j = fracView[rhFracOffset].pcrh_pfrhIdx();          // rechit index for this thread
+      rhPos = Position4{pfRecHits[j].x(), pfRecHits[j].y(), pfRecHits[j].z(), 1.};
+      rhEnergy = pfRecHits[j].energy();
+      rhPosNorm = fmaxf(0., logf(rhEnergy * rhENormInv));
+    }
+    alpaka::syncBlockThreads(acc);
+
+    do {
+      if (debug && tid == 0) {
+        printf("\n--- Now on iter %d for topoId %d ---\n", iter, topoId);
+      }
+      float dist2 = -1., d2 = -1., fraction = -1.;
+      if (tid < nRHOther) {
+        // Rechit distance calculation
+        dist2 = (clusterPos.x - rhPos.x) * (clusterPos.x - rhPos.x) +
+                (clusterPos.y - rhPos.y) * (clusterPos.y - rhPos.y) +
+                (clusterPos.z - rhPos.z) * (clusterPos.z - rhPos.z);
+
+        d2 = dist2 / pfClusParams.showerSigma2();
+        fraction = clusterEnergy * rhENormInv * expf(-0.5 * d2);
+
+        // For single seed clusters, rechit fraction is either 1 (100%) or -1 (not included)
+        if (fraction > pfClusParams.minFracTot() && d2 < 100.)
+          fraction = 1.;
+        else
+          fraction = -1.;
+        fracView[rhFracOffset].pcrh_frac() = fraction;
+      }
+      alpaka::syncBlockThreads(acc);
```

Comment:
I believe that this is guaranteed and a safe operation, since the Do While loop is based on a shared memory variable, similar to the earlier concern that was answered by @fwyzard.

Reply:
I'm wondering if it would be feasible to have a short comment for every `syncBlockThreads()` call that is inside a conditional block explaining why it is safe (i.e. either all or none of the threads of the block call the function)?

Reply:
I am happy to go through and add them